### DISCUSSION TITLE
Does `gtk_menu_popup_at_pointer()` work?

### DISCUSSION THREAD

**Discussion by DemiMarie** [wmww/gtk-layer-shell#193] [Q&A]
I’m trying to get Qubes OS’s widgets to work under Wayland.  They are implemented as `GtkStatusIcon`s that create `GtkMenu`s with `gtk_menu_popup_at_pointer()`.  Is this expected to work with Layer Shell, provided that an alternative app indicator implementation (such as Ayatana indicators) is used?   If not, could it be made to work reliably?

Under X11, they work fine, but with Xwayland they don’t because Xwayland doesn’t receive input events for parts of the screen that don’t belong to it.  I’m wondering if gtk-layer-shell could be used to port this application to native Wayland (under KDE or Sway).

**Comment by Consolatis** [comment]
I am using both, `gtk_menu_popup_at_widget()` and `gtk_menu_popup_at_pointer()` with a gtk 3 layershell panel and both work fine for me. For the `_at_pointer()` variant the cursor should be over a gtk surface belonging to the same application though.

**Reply by DemiMarie** [reply]
That’s an unfortunate restriction.  In my application (Wayland port of https://github.com/QubesOS/qubes-desktop-linux-manager), it is necessary to pop up a surface at a specific location in an output-local coordinate space.  Is that possible with this library, perhaps using the deprecated `gtk_menu_popup()` API?

**Reply by Consolatis** [reply]
> pop up a surface at a specific location in an output-local coordinate space. Is that possible with this library

Technically yes but its not the intended use-case for the layershell protocol. You could first get the output dimensions and then use [gtk_layer_set_margin](https://wmww.github.io/gtk-layer-shell/gtk-layer-shell-Gtk-Layer-Shell.html#gtk-layer-set-margin) to position it anywhere on the screen. That would then be a window though and not a menu. Although I guess you could try to spawn a menu from that  (potentially transparent) surface but this really is a hacky solution to get the layershell protocol to do something it wasn't designed for and goes somewhat against the design principles of wayland itself. I've also never tried it.

Depending on the actual use-case you could also spawn a fullscreen transparent layershell window (with all edge anchors set). That way you should also get mouse events but those would then obviously not get relayed anymore to whatever is under your transparent window. This is how tools like [slurp](https://github.com/emersion/slurp) work.

**Reply by DemiMarie** [reply]
Does this mean that e.g. an application launcher menu that drops down from a toolbar must be rendered in the toolbar process, rather than being an external program?  That’s going to be a serious problem for Qubes OS.

**Reply by Consolatis** [reply]
Yes. You usually can't embed external surfaces owned by another wayland client. There exists the concept of subsurfaces which could maybe be used but that is out of my personal experience. Another option might be to be your own sub compositor and the applets being wayland clients for your own compositor. I've never tried that either though. 

From what I remember xfce4-panel for example worked around that for out-of-process applets by sharing pixmap fds rendered by applets running as external processes over dbus to the main application which then creates windows out of them. I am not sure how they've dealt with external applets spawning menus, maybe via some dbusmenu variant. The easiest is indeed to just have the main process (or some thread) render everything itself.

**Reply by DemiMarie** [reply]
There is xdg-foreign-unstable-v2 which supports this but requires explicit cooperation between the processes.

**Reply by Consolatis** [reply]
That one only supports setting the parent to a surface of another process though. It also states
> Only xdg_toplevel equivalent surfaces may be exported, otherwise an invalid_surface protocol error is sent.

So I am not sure it would work for `xdg_popup`.
